hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyLifecycleService.java

```diff
@@ -366,171 +371,302 @@ private void evaluateFSOBucket(OmVolumeArgs volume, OmBucketInfo bucket, String
+                if (rule.getEffectivePrefix() != null && subDirPath.startsWith(rule.getEffectivePrefix())) {
+                  matched = true;
+                  break;
+                }
+              }
+              if (!matched) {
+                iterator.remove();
               }
-              expiredDirList.add(dirInfo.getName(), 0, dirInfo.getUpdateID());
             }
-          } else {
-            dirInfo = null;
-            LOG.info("Skip evaluate trash directory {}", TRASH_PREFIX);
+          }
+
+          if (!subDirSummary.getSubDirList().isEmpty()) {
+            item.setSubDirSummary(subDirSummary);
+            try {
+              stack.push(item);
+            } catch (CapacityFullException e) {
+              LOG.warn("Abort evaluate {}/{} at {}", volumeName, bucketName, currentDirPath, e);
+              return;
+            }
+
+            // depth first evaluation, push subDirs into stack
+            for (OmDirectoryInfo subDir : subDirSummary.getSubDirList()) {
+              String subDirPath = currentDirPath.isEmpty() ? subDir.getName() :
+                  currentDirPath + OM_KEY_PREFIX + subDir.getName();
+              try {
+                stack.push(new PendingEvaluateDirectory(subDir, subDirPath, null));
+              } catch (CapacityFullException e) {
+                LOG.warn("Abort evaluate {}/{} at {}", volumeName, bucketName, subDirPath, e);
+                return;
+              }
+            }
+            newSubDirPushed = true;
+          }
+        } else {
+          // this item is a parent directory, check how many sub directories are deleted.
+          for (OmDirectoryInfo subDir : subDirSummary.getSubDirList()) {
+            if (deletedDirSet.remove(subDir.getObjectID())) {
+              deletedDirCount++;
+            }
+          }
+        }
+
+        if (newSubDirPushed) {
+          continue;
+        }
+
+        // evaluate direct files, first check cache, then check table
+        // there are three cases:
+        // a. key is deleted in cache, while it's not deleted in table yet
+        // b. key is new added in cache, not in table yet
+        // c. key is updated in cache(rename), but not updated in table yet
+        //    in this case, the fromKey is a deleted key in cache, and the toKey is a newly added key in cache,
+        //    and fromKey is also in table
+        long numKeysUnderDir = 0;
+        long numKeysExpired = 0;
+        HashSet<String> deletedKeySet = new HashSet();
+        Iterator<Map.Entry<CacheKey<String>, CacheValue<OmKeyInfo>>> cacheIter = keyTable.cacheIterator();
+        while (cacheIter.hasNext()) {
+          Map.Entry<CacheKey<String>, CacheValue<OmKeyInfo>> entry = cacheIter.next();
+          OmKeyInfo key = entry.getValue().getCacheValue();
+          if (key == null) {
+            deletedKeySet.add(entry.getKey().getCacheKey());
+            continue;
+          }
+          numKeysUnderDir++;
```

Comment:
we should check whether this is current dir's file (`if (key.getParentObjectID() == currentDirObjID)`) before add the `numKeysUnderDir`